grype/pkg/syft_json_provider.go

```diff
@@ -20,33 +21,57 @@ import (
 // resilient to multiple syft JSON schemas (to a degree).
 // TODO: add version detection and multi-parser support (when needed in the future)
 func syftJSONProvider(config providerConfig) ([]Package, Context, error) {
-	var reader io.Reader
+	reader, err := getSyftJSON(config)
+	if err != nil {
+		return nil, Context{}, err
+	}
+
+	return parseSyftJSON(reader)
+}
+
+func getSyftJSON(config providerConfig) (io.Reader, error) {
 	if config.reader != nil {
-		// the caller is explicitly hinting to use the given reader as input
-		reader = config.reader
-	} else {
-		// try and get a reader from the description
-		if strings.HasPrefix(config.userInput, "sbom:") {
-			// the user has explicitly hinted this is an sbom, if there is an issue return the error
-			filepath := strings.TrimPrefix(config.userInput, "sbom:")
-			sbomReader, err := os.Open(filepath)
-			if err != nil {
-				return nil, Context{}, fmt.Errorf("user hinted 'sbom:' but couldn't read SBOM file: %w", err)
-			}
-			reader = sbomReader
-		}
+		// the caller has explicitly indicated to use the given reader as input
+		return config.reader, nil
+	}
 
-		// the user has not hinted that this may be a sbom, but lets try that first...
-		if sbomReader, err := os.Open(config.userInput); err != nil {
-			if errors.Is(err, os.ErrNotExist) {
-				return nil, Context{}, errDoesNotProvide
-			}
-		} else {
-			reader = sbomReader
+	if explicitlySpecifyingSBOM(config.userInput) {
+		filepath := strings.TrimPrefix(config.userInput, "sbom:")
+
+		sbom, err := openSbom(filepath)
+		if err != nil {
+			return nil, fmt.Errorf("unable to use specified SBOM: %w", err)
 		}
+
+		return sbom, nil
 	}
 
-	return parseSyftJSON(reader)
+	// as a last resort, see if the raw user input specified an SBOM file
+	sbom, err := openSbom(config.userInput)
+	if err == nil {
+		return sbom, nil
+	}
+
+	// no usable SBOM is available
+	return nil, errDoesNotProvide
+}
+
+func openSbom(path string) (*os.File, error) {
+	expandedPath, err := homedir.Expand(path)
```

Comment:
more of an open-ended question, but how does this expand when using `sudo`? is it acceptable to expand to `/root` instead of `/home/myusername` ?

Reply:
Interesting question... I would guess so. The homedir library tries its best to rely on the OS's own behavior for "home directory" lookup. (E.g. see here -> https://github.com/mitchellh/go-homedir/blob/af06845cf3004701891bf4fdb884bfe4920b3727/homedir.go#L89)

Reply:
This can be tricky in system calls using shell and interpreting back into code (I've run into issues like this in Python). Just tried it out with a quick test and it checks out. Thanks for the link the source code!